homeassistant/components/scheduler/schedule.py

```diff
@@ -0,0 +1,334 @@
+        )
+
+        await self._hass.services.async_call(
+            "scene",
+            "turn_on",
+            service_data={"entity_id": self.entity_id},
+            blocking=True,
+            context=self._context,
+        )
+
+        _LOGGER.info("Scheduler triggered scene: %s", self.entity_id)
+
+
+class Schedule:
+    """A class to represent a schedule."""
+
+    def __init__(
+        self,
+        hass: HomeAssistant,
+        entity_id: str,
+        start_datetime: datetime,
+        *,
+        end_datetime: Optional[datetime] = None,
+        recurrence: Optional[rrule] = None,
+    ):
+        """Initialize."""
+        self._async_revert_listener: Optional[Callable[..., Awaitable]] = None
+        self._async_trigger_listener: Optional[Callable[..., Awaitable]] = None
+        self._hass: HomeAssistant = hass
+        self._initial_instance_scheduled: bool = False
+        self.active_instance: Optional[ScheduleInstance] = None
+        self.end_datetime: Optional[datetime] = end_datetime
+        self.entity_id: str = entity_id
+        self.recurrence: Optional[rrule] = recurrence
+        self.schedule_id: str = uuid4().hex
+        self.start_datetime: datetime = start_datetime
+
+        self.instance_duration: Optional[timedelta]
+        if self.end_datetime:
+            self.instance_duration = self.end_datetime - self.start_datetime
+        else:
+            self.instance_duration = None
+
+    def __str__(self) -> str:
+        """Define the string representation of this schedule."""
+        return (
+            f'<Schedule start="{self.start_datetime}" '
+            f'end="{self.end_datetime}" rrule="{self.recurrence}">'
+        )
+
+    @property
+    def active(self) -> bool:
+        """Return whether the schedule has an active instance."""
+        # Instances with no end datetimes are one-shots and are not active:
+        if not self.end_datetime:
+            return False
+
+        now = dt_util.utcnow()
+
+        # If we're between the initial start and the end time, we're active:
+        if self.start_datetime <= now <= self.end_datetime:
+            return True
+
+        # If we have no recurrence (and we've already established that we're outside the
+        # initial start and end datetimes), we aren't active:
+        if not self.recurrence:
+            return False
+
+        # If there's no "before" datetime, that means we're at the beginning of the
+        # recurrence; since we've already established that there _is_ a recurrence,
+        # we're active:
+        last_start_dt = self.recurrence.before(now, inc=True)
+        if not last_start_dt:
+            return True
+
+        # Lastly, look at the recurrence just before right now and see
+        # if we're still inside it; if we are, we're active:
+        last_end_dt = last_start_dt + self.instance_duration
+        return last_start_dt <= now <= last_end_dt
+
+    @property
+    def expired(self) -> bool:
```

Comment:
Should we also introduce a `state` property that is either "on", "off" or "expired" ? The one that is "off" can still be turned on. "expired" cannot.

Reply:
I like it.